[ControlHeightReduction] Drop lifetime annotations where necessary

boomanaiden154 opened the pull request:
ControlHeightReduction will duplicate some blocks and insert phi nodes in exit blocks of regions that it operates on for any live values. This includes allocas. Having a lifetime annotation refer to a phi node was made illegal in 92c55a315eab455d5fed2625fe0f61f88cb25499, which causes the verifier to fail after CHR.

There are some cases where we might not need to drop lifetime annotations (usually because we do not need the phi to begin with), but drop all annotations for now to be conservative.

Fixes #159621.

boomanaiden154 committed:
[ControlHeightReduction] Drop lifetime annotations where necessary

ControlHeightReduction will duplicate some blocks and insert phi nodes
in exit blocks of regions that it operates on for any live values. This
includes allocas. Having a lifetime annotation refer to a phi node was
made illegal in 92c55a315eab455d5fed2625fe0f61f88cb25499, which causes
the verifier to fail after CHR.

There are some cases where we might not need to drop lifetime
annotations (usually because we do not need the phi to begin with), but
drop all annotations for now to be conservative. I believe this also
ends up being a correctness issue on top of the assertion because the
lifetime annotation in some cases would get shifted to after the store,
making the store UB when it was not before.

Fixes #159621.

nikic reviewed (commented):
I don't think this fix is quite right. For your example, CHR does something very undesirable, which is to split blocks in a way that ends up duplicating a previously static alloca, turning it into two dynamic allocas. We should really make sure that static allocas stay static. Additionally, I'm concerned that your approach here may end up partially removing lifetime markers. If we drop any markers, we always need to drop all of them.

The issue here looks very similar to the problem I fixed in https://github.com/llvm/llvm-project/commit/fc90685354ca36c935ebb0f315bb18592598a83b. I think you basically want to do the same. That is, for static allocas, make sure they are not part of the splitting. And for dynamic allocas, drop all lifetime markers if any have to be dropped.

boomanaiden154 committed:
feedback part1

boomanaiden154 committed:
feedback part2

boomanaiden154 committed:
code formatting

boomanaiden154 commented:
> I don't think this fix is quite right. For your example, CHR does something very undesirable, which is to split blocks in a way that ends up duplicating a previously static alloca, turning it into two dynamic allocas. We should really make sure that static allocas stay static. Additionally, I'm concerned that your approach here may end up partially removing lifetime markers. If we drop any markers, we always need to drop all of them.

I've updated CHR to hoist static allocas to the true entry block. In that case we also don't need to generate a phi node and we can preserve the lifetime annotations. I've added another test case for dynamic allocas where we do need to remove lifetime annotations and update `insertTrivialPHIs` to drop all lifetime annotations if we end up in a situation where we are trying to annotate the lifetime of a phi.

> The issue here looks very similar to the problem I fixed in https://github.com/llvm/llvm-project/commit/fc90685354ca36c935ebb0f315bb18592598a83b. I think you basically want to do the same. That is, for static allocas, make sure they are not part of the splitting. And for dynamic allocas, drop all lifetime markers if any have to be dropped.

Thanks for the example that was super helpful. And thank you for the clear explanation/your patience as I get a better grip on the middle end.

boomanaiden154 committed:
Merge branch 'main' into fix-chr-lifetimes

nikic reviewed (approved):
LGTM, thanks!
